src/servers/launcher/main.cpp

```diff
@@ -0,0 +1,849 @@
+
+      return std::array<const uint8_t, 16>{
+        uuid[0],
+        uuid[1],
+        uuid[2],
+        uuid[3],
+        uuid[4],
+        uuid[5],
+        uuid[6],
+        uuid[7],
+        uuid[8],
+        uuid[9],
+        uuid[10],
+        uuid[11],
+        uuid[12],
+        uuid[13],
+        uuid[14],
+        uuid[15],
+      };
+    }()) {}
+  static UUID unsafeCreate(const char* v) {
+    // No bounds checking, the caller is responsible for providing the
+    // right length array.
+    std::array<uint8_t, 16> uuid{};
+    for (size_t i{0}; i < 16; i++) uuid[i] = static_cast<uint8_t>(v[i]);
+
+    return UUID{uuid};
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const UUID& uuid) {
+    auto flags{os.flags()};
+    os << std::hex;
+    for (const auto& [i, c] : uuid.uuid | std::views::enumerate) {
+      os << std::setw(2) << std::setfill('0') << +c;
+      if (i == 3 || i == 5 || i == 7 || i == 9) os << '-';
+    }
+    os.flags(flags);
+
+    return os;
+  }
+};
+
+struct Context {
+  const UUID* SERVER_ID;
+};
+
+static Context ctx{};
+
+class DBConnectionError: public std::runtime_error {
+  public:
+  DBConnectionError(const std::string& error) noexcept:
+    std::runtime_error(error) {}
+};
+
+class DB {
+  PGconn* c;
+
+  DB() :
+    c([]{
+      PGconn* c{PQconnectdb("")};
+      if (PQstatus(c) != CONNECTION_OK) {
+        throw DBConnectionError(PQerrorMessage(c));
+      }
+      return c;
+    }()) {}
+
+  public:
+  static std::expected<DB, std::string> tryCreate() noexcept {
+    try {
+      return DB();
+    } catch (const DBConnectionError& exc) {
+      return std::unexpected(exc.what());
+    }
+  }
+  static const DB create() noexcept {
+    while (true) {
+      auto maybeDB{DB::tryCreate()};
+      if (maybeDB.has_value()) {
+        return std::move(maybeDB).value();
+      } else {
+        std::cout << maybeDB.error() << std::endl;
+        std::this_thread::sleep_for(RETRY_DELAY);
+      }
+    }
+  }
+  DB(const DB&) = delete;
+  DB(DB&& other) noexcept: c(std::exchange(other.c, nullptr)) {}
+  DB& operator=(const DB&) = delete;
+  DB& operator=(DB&& other) noexcept {
+    std::swap(c, other.c);
+
+    return *this;
+  }
+  ~DB() { PQfinish(c); }
+
+  const PGconn* getConn() const noexcept { return c; }
+};
+
+class QueryError: public std::runtime_error {
+  public:
+  QueryError(const std::string& error) noexcept:
+    std::runtime_error(error) {}
+};
+
+class Query {
+  PGresult* r;
+  Query(const PGconn* conn, const std::string& sql) :
+    r([&conn, &sql]{
+      PGresult* r{PQexecParams(
+        const_cast<PGconn*>(conn),
+        sql.c_str(),
+        0,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr,
+        1
+      )};
+      {
+        const ExecStatusType s{PQresultStatus(r)};
+        if (s != PGRES_TUPLES_OK && s != PGRES_COMMAND_OK) {
+          throw QueryError(PQerrorMessage(conn));
+        }
+      }
+
+      return r;
+    }()) {}
```

Comment:
Again, I’m not sure what the point is of writing the entire constructor as an IILE of one of the data members. There is no need to be so “clever” or circuitous.

But the really weird thing about this constructor is this:

```cpp
  Query(const PGconn* conn, const std::string& sql) :
    // ...
        const_cast<PGconn*>(conn),
```

??? *Why?* Why not just take `conn` as a `PGconn*` to begin with?